### DISCUSSION TITLE
Broken paginator ?

### DISCUSSION THREAD

**Discussion by piotrh052b** [filamentphp/filament#9957] [Help] [answered]
Hello,
After running the composer update command, I suddenly got an error message:

Declaration of Filament\Tables\Concerns\InteractsWithTable::paginateTableQuery(Illuminate\Database\Eloquent\Builder $query): Illuminate\Contracts\Pagination\Paginator|Illuminate\Contracts\Pagination\CursorPaginator must be compatible with Filament\Widgets\TableWidget::paginateTableQuery(Illuminate\Database\Eloquent\Builder $query): Illuminate\Contracts\Pagination\Paginator

at vendor\filament\tables\src\Concerns\CanPaginateRecords.php:28

Is there something wrong with my code or something is broken in filament v3.1.0 ?

**Comment by leandrocfe** [comment]
This version is working here.
Are you using a table widget?

**Reply by piotrh052b** [reply]
Yes and everything was working before composer update.

**Reply by leandrocfe** [reply]
Working here as well. Could you share the widget code please?

**Reply by piotrh052b** [reply]
This is my code

`<?php

namespace App\Filament\Resources\NoResource\Widgets;

use BezhanSalleh\FilamentShield\Traits\HasWidgetShield;
use Carbon\Carbon;
use Filament\Tables;
use Filament\Tables\Columns\IconColumn;
use Filament\Tables\Columns\TextColumn;
use Filament\Tables\Table;
use Filament\Widgets\TableWidget as BaseWidget;
use Illuminate\Contracts\View\Factory;
use Illuminate\Contracts\View\View;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Foundation\Application;
use Illuminate\Notifications\DatabaseNotification;
use Illuminate\Support\HtmlString;

class Notifications extends BaseWidget
{
    use HasWidgetShield;

    use Tables\Concerns\InteractsWithTable;

    protected int|string|array $columnSpan = 'full';

    public function table(Table $table): Table
    {
        return $table
            ->query(fn(): Builder => DatabaseNotification::query()
                ->where('notifiable_id', auth()->user()->id)
                ->take(5)
                ->latest()
            )
            ->recordClasses(function (DatabaseNotification $record) {
                if ($record->read_at) {
                    return 'bg-gray-100';
                }

                return '';
            })
            ->columns([
                IconColumn::make('notifiable_type')
                    ->label('')
                    ->icon(fn(DatabaseNotification $notification): string => $this->getNotificationIcon($notification)),
                TextColumn::make('type')
                    ->label('')
                    ->formatStateUsing(
                        fn(DatabaseNotification $notification): string => $this->getNotificationTitle($notification)
                    )
                    ->description(
                        fn(DatabaseNotification $notification): string => $this->getNotificationDescription($notification)
                    )
                    ->html(),
                TextColumn::make('id')
                    ->label('')
                    ->formatStateUsing(
                        fn(DatabaseNotification $notification): string => $this->getNotificationContent($notification)
                    )
                    ->html(),
            ])
            ->actions([
                Tables\Actions\Action::make('mark_read')
                    ->label('')
                    ->icon('heroicon-o-x-mark')
                    ->action(function (DatabaseNotification $notification) {
                        $notification->read_at = now();
                        $notification->save();
                    })
            ])
            ->headerActions([
                Tables\Actions\Action::make('mark_all_read')
                    ->label(__('Odznacz wszystkie jako przeczytane'))
                    ->action(fn() => $this->markAllRead()),
                Tables\Actions\Action::make('clear')
                    ->label(__('Wyczyść'))
                    ->modalHeading(__('Usuń wszystkie powiadomienia'))
                    ->requiresConfirmation()
                    ->action(fn() => $this->clearAll())
            ])
            ->heading($this->getHeading())
            ->paginated(false);
    }

    private function getNotificationTitle(DatabaseNotification $record): string
    {
        return new HtmlString('<span class="font-bold">' . $record->data['title'] . '</span>');
    }

    private function getNotificationDescription(DatabaseNotification $record): string
    {
        return Carbon::createFromFormat('Y-m-d H:i:s', $record->created_at)
            ->diffForHumans();
    }

    private function getHeading(): HtmlString
    {
        $count = DatabaseNotification::where('notifiable_id', auth()->user()->id)
            ->whereNull('read_at')
            ->count();

        return new HtmlString(view('tables.columns.notification.heading', ['count' => $count]));
    }

    public function getNotificationContent($payload): View|Application|Factory|\Illuminate\Contracts\Foundation\Application
    {
        $className = $payload->type;

        try {
            $obj = new \ReflectionClass($className);
            $instance = $obj->newInstanceWithoutConstructor();

            return $instance->contentView($payload->data['payload']);
        } catch (\ReflectionException $e) {
            logger($e->getMessage());
        }

        $view = 'tables.columns.notification.empty';

        return view($view);
    }

    private function markAllRead(): void
    {
        DatabaseNotification::where('notifiable_id', auth()->user()->id)
            ->update(['read_at' => now()]);

        $this->resetTable();
    }

    private function clearAll(): void
    {
        DatabaseNotification::where('notifiable_id', auth()->user()->id)
            ->delete();

        $this->resetTable();
    }

    private function getNotificationIcon(DatabaseNotification $notification)
    {
        return $notification->data['icon'] ?? 'heroicon-o-shopping-bag';
    }
}
`

**Reply by hugomyb** [reply, marked as answer]
Hi, I found the problem, in one of my TableWidget, I had a `use InteractsWithTable`, that's what was causing the problem. You don't need it, as the TableWidget already uses this trait, so you should try removing the `use InteractsWithTable` in your widget.

**Comment by JibayMcs** [comment]
HI !

Same problem here ! With a fresh install of the v3.1.0

```bash
=> composer install

Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Generating optimized autoload files
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover --ansi
PHP Fatal error:  Declaration of Filament\Tables\Concerns\InteractsWithTable::paginateTableQuery(Illuminate\Database\Eloquent\Builder $query): Illuminate\Contracts\Pagination\Paginator|Illuminate\Contracts\Pagination\CursorPaginator must be compatible with Filament\Widgets\TableWidget::paginateTableQuery(Illuminate\Database\Eloquent\Builder $query): Illuminate\Contracts\Pagination\Paginator in /home/jibaymcs/Programming/Programming/blouses-roses-v2/vendor/filament/tables/src/Concerns/CanPaginateRecords.php on line 28

   Symfony\Component\ErrorHandler\Error\FatalError 

  Declaration of Filament\Tables\Concerns\InteractsWithTable::paginateTableQuery(Illuminate\Database\Eloquent\Builder $query): Illuminate\Contracts\Pagination\Paginator|Illuminate\Contracts\Pagination\CursorPaginator must be compatible with Filament\Widgets\TableWidget::paginateTableQuery(Illuminate\Database\Eloquent\Builder $query): Illuminate\Contracts\Pagination\Paginator

  at vendor/filament/tables/src/Concerns/CanPaginateRecords.php:28
     24▕ 
     25▕         $this->resetPage();
     26▕     }
     27▕ 
  ➜  28▕     protected function paginateTableQuery(Builder $query): Paginator | CursorPaginator
     29▕     {
     30▕         $perPage = $this->getTableRecordsPerPage();
     31▕ 
     32▕         /** @var LengthAwarePaginator $records */


   Whoops\Exception\ErrorException 

  Declaration of Filament\Tables\Concerns\InteractsWithTable::paginateTableQuery(Illuminate\Database\Eloquent\Builder $query): Illuminate\Contracts\Pagination\Paginator|Illuminate\Contracts\Pagination\CursorPaginator must be compatible with Filament\Widgets\TableWidget::paginateTableQuery(Illuminate\Database\Eloquent\Builder $query): Illuminate\Contracts\Pagination\Paginator

  at vendor/filament/tables/src/Concerns/CanPaginateRecords.php:28
     24▕ 
     25▕         $this->resetPage();
     26▕     }
     27▕ 
  ➜  28▕     protected function paginateTableQuery(Builder $query): Paginator | CursorPaginator
     29▕     {
     30▕         $perPage = $this->getTableRecordsPerPage();
     31▕ 
     32▕         /** @var LengthAwarePaginator $records */

      +1 vendor frames 

  2   [internal]:0
      Whoops\Run::handleShutdown()

Script @php artisan package:discover --ansi handling the post-autoload-dump event returned with error code 255
```

**Comment by hugomyb** [comment]
Hi, I have exactly the same problem. 

Here is my code: 

```php
<?php

namespace App\Filament\Resources\ActivityLocationResource\Pages;

use App\Filament\Resources\ActivityLocationResource;
use App\Models\ActivityLocation;
use Closure;
use Filament\Tables\Actions\Action;
use Filament\Tables\Actions\CreateAction;
use Filament\Tables\Columns\TextColumn;
use Filament\Widgets\TableWidget as BaseWidget;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Relations\Relation;

class ListActivityLocations extends BaseWidget
{
    protected int|string|array $columnSpan = 'full';

    protected static ?string $heading = 'Lieux';

    protected function getTableQuery(): Builder|Relation|null
    {
        return ActivityLocation::query();
    }

    protected function getTableColumns(): array
    {
        return [
            TextColumn::make('name')
                ->label('Lieu')
                ->searchable()
                ->sortable(),

            TextColumn::make('committee.name')
                ->label('Comité')
                ->searchable()
                ->sortable(),
        ];
    }

    protected function getTableHeaderActions(): array
    {
        return [
            CreateAction::make('create_activity_location')
                ->url(ActivityLocationResource::getUrl('create'))
                ->label('Nouveau Lieu'),
        ];
    }

    protected function getTableActions(): array
    {
        return [
            Action::make('edit_activity_location')
                ->label('Modifier')
                ->icon('heroicon-m-pencil-square')
                ->url(fn(ActivityLocation $ActivityLocation): string => ActivityLocationResource::getUrl('edit', ['record' => $ActivityLocation])),
        ];
    }

    protected function getTableRecordUrlUsing(): ?Closure
    {
        return fn(ActivityLocation $ActivityLocation): string => ActivityLocationResource::getUrl('edit', ['record' => $ActivityLocation]);
    }
}
```
